File: bindings/javascript/samples/request_reply_async_sample.js
// SPDX-License-Identifier: MPL-2.0

'use strict';

const assert = require('node:assert/strict');
const { once } = require('node:events');
const net = require('node:net');
const zlink = require('@zlink-systems/zlink');
const { waitForConnectionReady } = require('./sample_support');

async function reservePort() {
  const srv = net.createServer();
  srv.listen(0, '127.0.0.1');
  await once(srv, 'listening');
  const { port } = srv.address();
  await new Promise((resolve, reject) => srv.close((error) => error ? reject(error) : resolve()));
  return port;
}

function timeoutPromise(ms, label) {
  return new Promise((_, reject) => {
    setTimeout(() => reject(new Error(`${label} timed out`)), ms);
  });
}

async function main() {
// --8<-- [start:doc]
  const port = await reservePort();
  const endpoint = `tcp://127.0.0.1:${port}`;
  const ctx = zlink.createContext();
  const routerSocket = zlink.createRouterSocket(ctx);
  const dealerSocket = zlink.createDealerSocket(ctx);
  const clientRoutingId = zlink.RoutingId.from(Buffer.from('request-reply-client'));

  try {
    const routerMonitor = routerSocket.monitorOpen([zlink.MonitorEventType.ConnectionReady]);
    const dealerMonitor = dealerSocket.monitorOpen([zlink.MonitorEventType.ConnectionReady]);
    try {
      dealerSocket.setRoutingId(clientRoutingId);
      routerSocket.bind(endpoint);
      dealerSocket.connect(endpoint);
      await waitForConnectionReady(routerMonitor);
      await waitForConnectionReady(dealerMonitor);
    } finally {
      routerMonitor.close();
      dealerMonitor.close();
    }

    const pendingReply = dealerSocket.request()
      .message(Buffer.from('ping'))
      .timeout(2000)
      .submit();
    const request = new zlink.Received();
    routerSocket.recv(request);
    try {
      assert.equal(request.routingId.toBytes().toString(), 'request-reply-client');
      assert.ok(typeof request.requestSeq === 'bigint');
      routerSocket.reply(request.routingId, request.requestSeq)
        .message(Buffer.from('pong'))
        .submit();
    } finally {
      request.close();
    }
    const reply = await pendingReply;
    try {
      assert.equal(reply[0].data().toString(), 'pong');
    } finally {
      for (const part of reply) {
        part.close();
      }
    }
    console.log('[dealer-router/request-reply/async] send: "ping" -> recv: "pong"');
  } finally {
    dealerSocket.close();
    routerSocket.close();
    ctx.close();
  }
// --8<-- [end:doc]
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});

File: bindings/python/samples/request_reply_async_sample.py
import asyncio

import zlink

from sample_support import tcp_endpoint, wait_connected


def respond(router_socket):
    received = zlink.create_received()
    if not router_socket.recv_into(received):
        raise AssertionError("expected request payload")
    with received:
        if received.routing_id != zlink.RoutingId(b"REQ-CLIENT"):
            raise AssertionError("unexpected request routing id")
        if received.reply_token is None:
            raise AssertionError("missing reply token")
        router_socket.reply(received.routing_id, received.reply_token).message(
            b"pong"
        ).submit()


async def main():
# --8<-- [start:doc]
    _, endpoint = tcp_endpoint()

    with zlink.create_context() as ctx:
        with zlink.create_router_socket(ctx) as router_socket:
            with zlink.create_dealer_socket(ctx) as dealer_socket:
                with router_socket.monitor_open(
                    zlink.MonitorEventMask.CONNECTION_READY
                ) as router_monitor:
                    with dealer_socket.monitor_open(
                        zlink.MonitorEventMask.CONNECTION_READY
                    ) as dealer_monitor:
                        dealer_socket.set_routing_id(b"REQ-CLIENT")
                        router_socket.bind(endpoint)
                        dealer_socket.connect(endpoint)
                        wait_connected(router_monitor, dealer_monitor)

                pending_reply = asyncio.create_task(
                    dealer_socket.request().message(b"ping").timeout(2.0).submit()
                )
                await asyncio.to_thread(respond, router_socket)
                reply = await pending_reply
                try:
                    if [part.to_bytes() for part in reply] != [b"pong"]:
                        raise AssertionError("unexpected reply payload")
                finally:
                    for part in reply:
                        part.close()
                print('[dealer-router/request-reply/async] send: "ping" -> recv: "pong"')
# --8<-- [end:doc]


if __name__ == "__main__":
    asyncio.run(main())
